### PR TITLE
Fixes #28879 - add pulp3 repo migration

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
TODO:
* [ ] resolve TODOs 
* [ ] write a vcr test
* [ ] figure out a way of showing progress

**Comment by jlsherrill** [comment]
I still have some more testing to do, but i think its generally ready for review

**Comment by jlsherrill** [comment]
docker migration is current blocked by: https://pulp.plan.io/issues/6097

**Comment by jlsherrill** [comment]
Testing tips:

1)  disable file/docker from pulp3 in katello.yaml:
```
:katello:
 
<OTHER STUFF>
 
  :use_pulp_2_for_content_type:
    :docker: true
    :file: true
```
2) Populate file and docker repos, promote them to content views, upload file content, etc...  Go crazy :)
3) run  'rake katello:pulp3_migration'
4) All the _href's on Repository objects should be filed out, Katello::Pulp3::RepositoryReferences should be created out, Katello::Pulp3::DistributionReferences should be created.  Verify as much as you can on the rails console
5) flip those two content types to pulp3 in katello.yaml, restart your rails server
6) verify you can access content, docker pull, upload new content, docker pull, promote repos in content views, publish new content views, etc...

**Comment by jturel** [comment]
low prio: is this needed for 3.15? Since we branched - trying to determine if it'll need to be cherry-picked.

**Comment by sjha4** [comment]
Still testing this. 
One point around the large file migration where for me, after migration there were:
[1] pry(main)> Katello::FileUnit.where(migrated_pulp3_href: nil).count
=> 65107
Pulp team says they haven't been able to reproduce this on later versions and given we are still on pulp_2to3_migration_client-0.0.1a1.dev01573066581, I think the issue will need to be retested after updating the client.

File: katello.gemspec [hunk 1]
```diff
@@ -49,7 +49,7 @@ Gem::Specification.new do |gem|
   gem.add_dependency "pulp_ansible_client", "<= 0.2.0b7.dev01574717759"
   gem.add_dependency "pulp_container_client", "<= 1.1.0.dev01574357179"
   gem.add_dependency "pulp_rpm_client", "<= 3.1.0b1.dev01576187357"
-  gem.add_dependency "pulp_2to3_migration_client", "<= 0.0.1a1.dev01573066581"
+  gem.add_dependency "pulp_2to3_migration_client", ">= 0.0.1b2.dev01580407602", "< 1.0"
```
Comment (sjha4) [inline katello.gemspec]: Aah..I missed this update..

**Comment by sjha4** [comment]
Testing with files. Results look good to me.
1. Migration Plan correctly formed: https://codebeautify.org/jsonviewer/cba546f1 :heavy_check_mark: 
2. Library repos are created with remotes (Map to importers) :heavy_check_mark:
3. Library and CV repos get distributions.(https://codebeautify.org/jsonviewer/cb858151) :heavy_check_mark:
4. Archived Repos do not get distributions created. :heavy_check_mark:
4. Publications are created and shared by distributions of the same archived repo. (https://codebeautify.org/jsonviewer/cba904a2) :heavy_check_mark:
5. All content units are copied. Except for large repo tests as in comment above. :heavy_check_mark:

Will test this a little more and also test with docker content.

**Comment by sjha4** [comment]
**Minimal Setup:**
CV published and promoted to dev environement with 1 docker repo:
1. Library version of docker repo
2. Archived version of docker CV version 1
3. Cv repos in library and dev environment
*****
**Upon migration:**
Migration Plan created. https://codebeautify.org/jsonviewer/cb4fd0d9
Docker repos in pulp3: https://codebeautify.org/jsonviewer/cb9739fb
2 repos : 1 corresponding to the original repo, and another for the archived repo
2 Publications corresponding the above 2 repos: https://codebeautify.org/jsonviewer/cbd8bd8f
3 total distributions: https://codebeautify.org/jsonviewer/cb4736d1
 - 2 distributions corresponsing to the archived repo distributed for Library and dev env.
 - 1 distribution for the repo in the default view.

**Comment by sjha4** [comment]
Able to see repo and distribution references getting created. :heavy_check_mark: 
Able to upload/deleted content from migrated repos in pulp3.. :heavy_check_mark: 
In case not switching to pulp3 after running migration, able to upload/delete in pulp2 and re-run migration and pulp3 repo gets a new version created to reflect changes made in pulp2.  :heavy_check_mark: 

1 issue I noticed was with deleting Repositories after migration and remaining on pulp2.
If a migrated repository is deleted, the references in katello are deleted along with remote, distribution, publication in pulp3. However, the repository itself and contents are not deleted in pulp3 even though the repo is deleted in pulp2.

**Comment by jlsherrill** [comment]
@sjha4 yes, that is a good point, we likely need an orphan cleanup task to run to clean those up.   I opened this to track it: https://projects.theforeman.org/issues/29079

**Comment by jlsherrill** [comment]
I also opened https://projects.theforeman.org/issues/29080 to allow for better progress reporting

**Comment by johnpmitsch** [comment]
> 1 issue I noticed was with deleting Repositories after migration and remaining on pulp2.
> If a migrated repository is deleted, the references in katello are deleted along with remote, distribution, publication in pulp3. However, the repository itself and contents are not deleted in pulp3 even though the repo is deleted in pulp2.

@jlsherrill I'm seeing the same thing, also the content views and content view versions seem to be still around in pulp2 and pulp3 after being deleted. I also ran `delete_orphan_contents` rake task

After the deletion:
```
[1] pry(main)> Katello::Pulp3::DistributionReference.all
=> []
[2] pry(main)> Katello::Pulp3::RepositoryReference.all
=> []
```
No content views (all versions and CVs deleted)
Screenshot from 2020-02-20 09-16-58

All repositories are deleted as well

Seeing this still in pulp3
```
[root@centos7-katello-devel-stable ~]# http GET http://admin:password@localhost:24817/pulp/api/v3/repositories/file/file/
HTTP/1.1 200 OK
Allow: GET, POST, HEAD, OPTIONS
Connection: close
Content-Length: 697
Content-Type: application/json
Date: Thu, 20 Feb 2020 14:19:54 GMT
Server: gunicorn/20.0.4
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

{
    "count": 2, 
    "next": null, 
    "previous": null, 
    "results": [
        {
            "description": "cv1-1-2", 
            "latest_version_href": null, 
            "name": "cv1-1-2", 
            "pulp_created": "2020-02-19T15:22:59.310798Z", 
            "pulp_href": "/pulp/api/v3/repositories/file/file/5cd65d85-b996-4b59-8ad3-5e5e35bc5d40/", 
            "versions_href": "/pulp/api/v3/repositories/file/file/5cd65d85-b996-4b59-8ad3-5e5e35bc5d40/versions/"
        }, 
        {
            "description": null, 
            "latest_version_href": null, 
            "name": "1-cv1-Library-73c9f2c5-f56d-44db-aaea-650de1bbffbf", 
            "pulp_created": "2020-02-19T02:06:07.008463Z", 
            "pulp_href": "/pulp/api/v3/repositories/file/file/9ecd9a33-0a83-4d9a-b6f7-6ca76bd8e595/", 
            "versions_href": "/pulp/api/v3/repositories/file/file/9ecd9a33-0a83-4d9a-b6f7-6ca76bd8e595/versions/"
        }
    ]
}
```

and still seeing CVs and repos in the pulp 2 repos
```

[root@centos7-katello-devel-stable ~]# pulp-admin -u admin -p $pulpAdminPassword repo list
+----------------------------------------------------------------------+
                              Repositories
+----------------------------------------------------------------------+

Id:                  73c9f2c5-f56d-44db-aaea-650de1bbffbf
Display Name:        file1
Description:         None
Content Unit Counts:
  Iso: 1

Id:                  2e90c5ab-a4c4-4a0c-95a9-acd8a7db7af5
Display Name:        docker1
Description:         None
Content Unit Counts:
  Docker Blob:     5
  Docker Manifest: 2
  Docker Tag:      2

Id:                  1-cv1-v1_0-2e90c5ab-a4c4-4a0c-95a9-acd8a7db7af5
Display Name:        docker1
Description:         None
Content Unit Counts:
  Docker Blob:     5
  Docker Manifest: 2
  Docker Tag:      2

Id:                  1-cv1-v1_0-73c9f2c5-f56d-44db-aaea-650de1bbffbf
Display Name:        file1
Description:         None
Content Unit Counts:
  Iso: 1

Id:                  1-cv1-Library-73c9f2c5-f56d-44db-aaea-650de1bbffbf
Display Name:        file1
Description:         None
Content Unit Counts:
  Iso: 1

Id:                  1-cv1-library-2e90c5ab-a4c4-4a0c-95a9-acd8a7db7af5
Display Name:        docker1
Description:         None
Content Unit Counts:
  Docker Blob:     5
  Docker Manifest: 2
  Docker Tag:      2
```

Sounds like your filed issue will take care of a lot of this, but just wanted to point out the CV/CVV deletion behavior, it seems like they should have been removed from pulp3 as well.

**Comment by johnpmitsch** [comment]
> Sounds like your filed issue will take care of a lot of this, but just wanted to point out the CV/CVV deletion behavior, it seems like they should have been removed from pulp3 as well.

We discussed this off-thread. The behavior above is mostly expected but there are two things to consider/look into:

- When you delete the full Content View, are the relevant pulp3 repositories deleted?
- What is the expectation for users for deleting pulp2 content after switching over to pulp3? Should it happen automatically or manually by the user? Should we handle for each content type or handle at the end of a full pulp3 switchover for all content types?

Not necessarily anything we have to solve here, but some questions raised by the PR.

**Comment by jlsherrill** [comment]
Opened:   https://projects.theforeman.org/issues/29122  for cv deletion
and https://projects.theforeman.org/issues/29124 for repo deletion after migration

**Comment by jlsherrill** [comment]
@sjha4 @johnpmitsch  ya'll good with this pr?

**Comment by jlsherrill** [comment]
thanks guys!!  I added a card for more migration testing

**Pull request merged**